include/alpaka/atomic/AtomicStdLibLock.hpp

```diff
@@ -81,22 +83,40 @@ namespace alpaka
         template<typename TOp, typename T, typename THierarchy, size_t THashTableSize>
         struct AtomicOp<TOp, AtomicStdLibLock<THashTableSize>, T, THierarchy>
         {
+            static constexpr bool IsLockFree
+                = std::is_trivially_copyable_v<T> && boost::atomic_ref<T>::required_alignment <= alignof(T)
```

Comment:
This compile-time check will never fail, because `boost::atomic_ref<T>::required_alignment` is probably just `alignof(T)`.

`boost::atomic_ref<T>::required_alignment` is a requirement on `addr`, i.e. would need to be checked at runtime.

Reply:
@jkelling wrote:
> This compile-time check will never fail, because `boost::atomic_ref<T>::required_alignment` is probably just `alignof(T)`.

According to the [Boost documentation](https://www.boost.org/doc/libs/1_77_0/doc/html/atomic/interface.html#atomic.interface.interface_atomic_ref) it is not necessarily the case:

> The referenced object must have alignment not less than indicated by the `boost::atomic_ref<T>::required_alignment` constant. That constant may be larger than the natural alignment of type `T`.

There are 9 files with over 9k lines of code inside Boost that specialise the atomic traits depending on the architecture, CPU capabilities, compiler, and actual type size - so I would not just assume that this is simply the same as `alignof(T)`.


@jkelling wrote:
> `boost::atomic_ref<T>::required_alignment` is a requirement on `addr`, i.e. would need to be checked at runtime.

We can assume that a variable of type `T` is aligned at least as much as `alignof(T)`, so if the compile time check succeeds, we don't need a runtime check.

What one could do is to add a second branch with a runtime check for the case where the compile time check fails (_i.e._ `boost::atomic_ref<T>::required_alignment` is larger than `alignof(T)`) but the actual variable is still suitably aligned (`addr` is a multiple of `boost::atomic_ref<T>::required_alignment`).

That could indeed be a further optimisation if one encounters cases where the compile time check fails.

Reply:
The implementation checks the alignment at runtime using a `BOOST_ASSERT` (So only in debug mode, I think?).

While it is good to check as much as possible at compile time. This check for alignment is not reliable, because at runtime a value may be aligned to higher or lower moduli than `alignof()` (the latter should only happen if someone is doing something bad, so this is not that important). Maybe it is best to just let the backend, i.e. boost, do the check, even though this would remove the fallback to `AtomicStdLibLock` and thus just fail in these corner cases.

The boost doc says `required_alignment` may be larger than the native alignment for lock-free atomics, for types on which atomics are not lock-free it is always equal to the native alignment. It looks like there is no fallback to a locking atomic is `alignof(T) <= alignment(value) < required_alignment`. This would probably be more efficient, than using Alpaka's `AtomicStdLibLock` in this case, because there the lock is shared between all operations at any address at level of atomicity.

Reply:
OK, we will rework this PR and write a complete replacement for `AtomicStdLibLock` (_e.g._ `AtomicBoostAtomicRef`) instead the current intrusive changes.

I would still prefer to keep
```c++
static_assert(std::is_trivially_copyable_v<T>, "Atomic operations can only be performed on variables of a trivially-copyable type.");
```
and
```c++
static_assert(boost::atomic_ref<T>::required_alignment <= alignof(T), "Atomic operations can only be performed variables of a correctly aligned type (see the documentation of boost.atomic_ref).");
```
On the other hand
  - if the `boost::atomic_ref<T>::required_alignment` does happen to be larger than `alignof(T)`, I cannot think of any workarounds, since I don't think it is possible to implement a properly-aligned type.
  - in my tests,  `boost::atomic_ref<T>::required_alignment`  was always the same as `alignof(T)`

So I'd be OK with keeping only the `trivially_copyable` check, and deferring the alignment check at runtime.

Reply:
> OK, we will rework this PR and write a complete replacement for `AtomicStdLibLock` (_e.g._ `AtomicBoostAtomicRef`) instead the current intrusive 

I left the `Boost` part out intentionally, so we can migrate it to STL without affecting anything using this implementation when alpaka moves to C++20 . Calling it `Std` would currently be wrong, hence just `AtomicAtomicRef`. How do others feel about this?

> I would still prefer to keep [... compile time alignment check]

Including `AtomicStdLibLock` as a fallback for that `alignof(T) < required_alignment<T>` case would entail keeping around that hash table of mutexes. This looks like too much cost for a case which you did encounter in 9k lines on your probably quite standard CPUs. Maybe this case does not occur for any alpaka users.

I am neutral on the compile-time check though, its negative impact is necessarily as rare as the benefit of the fallback and it avoids some, equally rare, runtime errors.

Reply:
I did not suggest keeping the `AtomicStdLibLock` fallback, but only having a `static_assert()` on the condition.

This would still have the benefit of catching the (rare, as you imply) occurrences where `alignof(T) < required_alignment<T>`, failing with a compile time error instead of a runtime error, without any extra cost.
No fallback would be provided (apart from the user changing the atomic implementation back to `AtomicStdLibLock` by hand).

> This looks like too much cost for a case which you did encounter in 9k lines 

I'm not sure if you actually read my comments ?
9k lines of code is what Boost needs to implement all the checks and traits used by `atomic_ref` for various architectures.
I certainly did not read through all of them - I simply trust the Boost documentation when it says that an explicit check is needed before using a certain feature, rather than assuming that it is probably just not needed or that maybe it does not occur.

> on your probably quite standard CPUs. Maybe this case does not occur for any alpaka users.

I have no objections to removing the check altogether, if you can guarantee that it can never happen on the architectures we are interested in: x86_64, aarch64, ppc64.

Reply:
> I did not suggest keeping the `AtomicStdLibLock` fallback, but only having a `static_assert()` on the condition.

It came up in both our comments, so I wanted to address it. After I wrote I realized, though, that this table is static, so the cost would be negligible when not used, so maybe one could do it after all. Either the way you did it already using `if consexpr` or using SFINAE to have `AtomicBoostAtomicRef` become `AtomicStdLibLock` in that case.

> This would still have the benefit of catching the (rare, as you imply) occurrences where `alignof(T) < required_alignment<T>`, failing with a compile time error instead of a runtime error, without any extra cost. No fallback would be provided (apart from the user changing the atomic implementation back to `AtomicStdLibLock` by hand).

That is what I wrote. I do not object to this option.

> > This looks like too much cost for a case which you did encounter in 9k lines
> 
> I'm not sure if you actually read my comments ? 9k lines of code is what Boost needs to implement all the checks and traits used by `atomic_ref` for various architectures. I certainly did not read through all of them - I simply trust the Boost documentation when it says that an explicit check is needed before using a certain feature, rather than assuming that it is probably just not needed or that maybe the problem does not occur.

I read them, but wrote a long time after reading. You mentioned, that you did not encounter it in your code and I thought your wrote something about the size of that, so I scrolled up and saw the 9k, sorry. The number does not actually matter though. I am also not saying the boost implementation is not needed. The have support for lots of platforms in there. The questions is just how to handle this in alpaka, because it does affect portability on the one had (i.e. when we do not check) but it could also break existing code if we check and do not have a fall back. The latter Could only be dismissed by assuming, that alpaka code is not running on any of those platforms and we do not care about them. Whether we want the fallback also depends on how much we still care.

> > on your probably quite standard CPUs. Maybe this case does not occur for any alpaka users.
> 
> I have no objections to removing the check altogether, if you can guarantee that it can never happen on the architectures we are interested in: x86_64, aarch64, ppc64.

Did you test this PR on all three already?